src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,682 @@
+  :: forall store pm. (StorageC store pm)
+  => Config pm -> '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+checkProposalLimitReached Config{..} = do
+  duupX @2
+  stToField #sProposalKeyListSortByDate
+  size
+  stackType @[Natural, ProposeParams pm, store]
+  toNamed #currentAmount
+  push (cMaxProposals); toNamed #maxAmount
+  if #maxAmount <=. #currentAmount then
+    failCustom_ #mAX_PROPOSALS_REACHED
+  else nop
+
+-- | Freeze the account's unfrozen token associated with the address
+freeze
+  :: forall store pm s. (StorageC store pm)
+  => Natural : Address : store : s :-> store : s
+freeze = do
+  dig @2
+  duupX @3
+  duupX @3
+  stackType @(Natural : Address : store : Natural : Address : s)
+
+  push unfrozenTokenId
+  pair
+  dip swap; swap; dip swap
+  stackType @(store : Address : (TokenId, Natural) : Natural : Address : s)
+  debitFrom
+
+  stackType @(store : Natural : Address : s)
+  dip $ do
+    push frozenTokenId
+    pair
+    swap
+
+  stackType @(store : Address : (TokenId, Natural) : s)
+  creditTo
+
+-- | Unfreeze the account's frozen token associated with the address
+unfreeze
+  :: forall store pm s. (StorageC store pm)
+  => Natural : Address : store : s :-> store : s
+unfreeze = do
+  dig @2
+  duupX @3
+  duupX @3
+  stackType @(Natural : Address : store : Natural : Address : s)
+
+  push frozenTokenId
+  pair
+  dip swap; swap; dip swap
+  stackType @(store : Address : (TokenId, Natural) : Natural : Address : s)
+  debitFrom
+
+  stackType @(store : Natural : Address : s)
+  dip $ do
+    push unfrozenTokenId
+    pair
+    swap
+
+  stackType @(store : Address : (TokenId, Natural) : s)
+  creditTo
+
+addProposal
+  :: forall store pm. (NicePackedValue pm, StorageC store pm)
+  => '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+addProposal = do
+  ensureProposalIsUnique
+  constructT @(Proposal pm)
+    ( fieldCtor $ push (0 :: Natural) -- initial upvote
+    , fieldCtor $ push (0 :: Natural) -- initial downvote
+    , fieldCtor $ now -- pStartDate
+    , fieldCtor $ getField #ppProposalMetadata
+    , fieldCtor $ sender
+    , fieldCtor $ getField #ppFrozenToken
+    , fieldCtor $ push []
+    )
+  dip swap
+  stackType @[Proposal pm, store, ProposeParams pm]
+  some
+  duupX @3; sender; swap; pair; pack @(ProposeParams pm, Address)
+  -- stToField #sProposalIdCounter
+  stackType @[ProposalKey, Maybe (Proposal pm), store, ProposeParams pm]
+  stUpdate #sProposals
+
+  stGetField #sProposalKeyListSortByDate
+  duupX @3; sender; swap; pair; pack @(ProposeParams pm, Address)
+  stackType @[ProposalKey, [ProposalKey], store, ProposeParams pm]
+  cons -- Assuming passed proposal start immediately
+  stackType @[[ProposalKey], store, ProposeParams pm]
+  stSetField #sProposalKeyListSortByDate
+
+  swap
+  -- incrementProposalId
+
+-- incrementProposalId
+--   :: forall store pm. (StorageC store pm)
+--   => '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+-- incrementProposalId = do
```

Comment:
Probably this thing can be removed now as well.

Reply:
Removed